### ISSUE TITLE
keras-tuner and MLFlow integration

### ISSUE THREAD

**Issue by frfritz** [keras-team/keras-tuner#494]
Is there going to be an option to track hyperparameter tuning with keras-tuner by mlflow? Not sure how that would be implemented, but I guess by either enabling custom objective functions or implementing a callback which handles mlflow.

**Comment by jonnor** [comment]
The KerasTuner API has an integration point designed for things like that. One can pass a Logger instance when setting up the search. Unfortunately it does not seem to be documented much. But you can see the interface here: https://github.com/keras-team/keras-tuner/blob/master/keras_tuner/engine/logger.py#L16

Create a sublass of that, and call mlflow from there. Example for logging the selected hyperparameters:
```
    def register_trial(self, trial_id, trial_state):
        hyperparams = trial_state['hyperparameters']['values']
        mlflow.log_params(hyperparams)
```

**Comment by fzyzcjy** [comment]
hi is there any updates? thanks!

**Comment by jonnor** [comment]
Here is the code I made to integrate mlflow and keras-tuner:
https://gist.github.com/jonnor/c107f3ca24a36c91d8ff94029a0cd357

I think that the code in LoggerTunerMixin should be added into keras-tuner (improvements to logging hooks). And the MlflowLogger into mlflow.

**Comment by abdulbasitds** [comment]
@jonnor, what does this do? does this display hyperparameter choices in UI?

**Comment by jonnor** [comment]
@abdulbasitds yes, it tracks parameters and metrics. It uses nested mlflow runs to keep track of each kearas tuner search, each trial in the search, and each execution of each trial. Unfortunately, the UI does not show the nesting so clearly, but it is in the data structure at least.
For now, there tracking of metrics and parameters is on the execution (each training run) - it is not merged up to the trial or the search yet.

See this screenshot for an example:
![keras-tuner-mlflow](https://user-images.githubusercontent.com/45185/144497146-b2c6747d-40a5-41ce-a8ae-2a5ca9fbe00a.png)

**Comment by abdulbasitds** [comment]
> @abdulbasitds yes, it tracks parameters and metrics. It uses nested mlflow runs to keep track of each kearas tuner search, each trial in the search, and each execution of each trial. Unfortunately, the UI does not show the nesting so clearly, but it is in the data structure at least. For now, there tracking of metrics and parameters is on the execution (each training run) - it is not merged up to the trial or the search yet.
> 
> See this screenshot for an example: ![keras-tuner-mlflow](https://user-images.githubusercontent.com/45185/144497146-b2c6747d-40a5-41ce-a8ae-2a5ca9fbe00a.png)

Thanks @jonnor , thanks

**Comment by rakshithvsk** [comment]
Hi @jonnor, This is great!! It is in fact a good work around to capture metrics per trial. 

However I noticed a major issue. If my epochs per trial is >1, ideally it should be persisting the best epoch out of all the epochs within a trial. But looks like its persisting the latest epoch per trial, though it not the best one. By any chance did you come across the same issue?
Any suggestions would help. Thanks!!

**Comment by jonnor** [comment]
Hi @rakshithvsk, that could very well be. I do not have time right now to have a look at it. If you find a fix, please do post the updated code as Gist :)

**Comment by NusretOzates** [comment]
Hey everyone, maybe [this approach ](https://towardsdev.com/using-mlflow-with-keras-tuner-f6df5dd634bc) could help.

I wrote the article while trying to do same thing
